docs/algorithms/nautilus/algorithm_en.md

```diff
@@ -0,0 +1,290 @@
+You can write the binary data in the array format like the following in the Nautilus mode of fuzzuf.
+```json
+["NULL", [0]]
+```
+You can also write a sequence of binary data. (Although JSON only supports decimal values unfortunately.)
+```json
+["DEADBEEF", [239, 190, 173, 222]]
+```
+If the array of binary data contains string, they will be concatenated and recognized as a single rule. In the following rules, for instance, the non-terminal `A` will be expanded to `\x00A\x00BBBB\xFFA\xFF`.
+```json
+["A", [0, "A{B}A", 255]],
+["B", [0, "BBBB", 255]],
+```
+
+You can also use the binary format when you want to merge multiple rules by `|` symbol. That is to say,
+```json
+["A", ["hello", [0], ["bye", 128]]]
+```
+and
+```json
+["A", "hello"],
+["A", [0]],
+["A", ["bye", 128]]
+```
+are equivalent.
+
+### 2-2. Testing Grammar Files
+You may want to check if your grammar is correct when it gets complicated.
+`tools/nautilus/generator` is the program to generate some random test cases according to the grammar you define. If you give it the grammar we defined previously, a random string will be generated like the following if the grammar is correct.
+```
+$ tools/nautilus/generator -g ./calc_grammar.json -t 100
+4/(((4)+1+7+7+5*4+1)-2-6+2-48-9*+5+52-6)
+$ tools/nautilus/generator -g ./calc_grammar.json -t 100
+45-((2))+8*-+9+2/4+7+4-(((3-6)/5)-2*9+(3)-7)
+```
+The following options are available, which you can also check by `--help` option.
+
+- `--grammar_path` / `-g`: Path to the grammar file [**Required**]
+- `--tree_depth` / `-t`: Maximum size of tree (The bigger this value is, the longer the output will be) [**Required**]
+- `--number_of_trees` / `-n`: Number of test cases to generate [Default:1]
+- `--store` / `-s`: Path to the folder to save generated test cases in [Default:None]
+
+If your grammar file is invalid, it will dump an error. The following error message will be shown if JSON is invalid.
+```
+[-] Cannot parse grammar file
+[json.exception.parse_error.101] parse error at line 3, column 5: syntax error while parsing array - unexpected '['; expected ']'
+```
+In this case, the 3rd line of the grammar file has invalid JSON format.
+
+A longer error like the following may also happen.
+```
+Found unproductive rules: (missing base/non recursive case?)
+START => EXPRESSION
+EXPRESSION => TERM
+EXPRESSION => EXPRESSION, +, EXPRESSION
+EXPRESSION => EXPRESSION, -, EXPRESSION
+TERM => FACTOR
+TERM => FACTOR, *, FACTOR
+TERM => FACTOR, /, FACTOR
+FACTOR => NUMBER
+FACTOR => (, EXPRESSION, )
+NUMBER => FACTOR
+terminate called after throwing an instance of 'exceptions::fuzzuf_runtime_error'
+  what():  Broken grammar
+```
+This error occurs if a non-terminal cannot reach the terminal. In the example above, `NUMBER => FACTOR` is invalid because it cannot reach the terminal by traversing `FACTOR`, which is a infinite loop.
```

Comment:
```suggestion
This error occurs if a non-terminal cannot reach the terminal. In the example above, `NUMBER => FACTOR` is invalid because it cannot reach the terminal by traversing `FACTOR`, which is an infinite loop.
```